### PR TITLE
highlighting: add support for simulating highlighting timeouts

### PR DESCRIPTION
**Pull request by slimsag**
This is useful for confirming our behavior of highlighting timeouts, and can be
used in the future to automate testing of this behavior as part of an e2e test.

Primarily adding to reproduce https://github.com/sourcegraph/sourcegraph/issues/4267

Note: This matches other cases where e.g. github.com/sourcegraphtest/AlwaysCloningTest
is a special repository for testing gitserver cloning states end-to-end.

To use this you just need to add github.com/sourcegraph/AlwaysHighlightTimeoutTest to your instance and view a file in there or search results coming from that repository.

Helps #4267

### CODE REVIEW

**Comment by slimsag** [comment]
You can see the current timeout right above the line in this PR where we are simulating it: https://github.com/sourcegraph/sourcegraph/pull/4364/files#diff-5a881b36b33270d5031b9a3060269516R49

Simulating with a large file is not a good idea because it would be prone to changing as syntect becomes faster and also would not cover the case of syntect hitting a bad language grammar definition.

**Comment by felixfbecker** [comment]
Can't we use a file that is so enormous it will _for sure_ hit a timeout? I think 3s is very reasonable for an e2e test.

**Comment by felixfbecker** [comment]
How does a simulated timeout test a bad language grammar?

**Comment by slimsag** [comment]
There are multiple failure cases to test here:

1. A super large file that syntect can't handle. This will **not** result in a contextual timeout and would instead result instead in syntect rejecting the request which ends up as an HTTP status code error and hits a different error path (but with the same end result).
2. syntect_server being unresponsive in some way, e.g. due to it being overwhelmed due to many concurrent requests or hitting a bad language grammar edge case that syntect can't handle. This is the cause of #4267 and the only way to test this is via (1) a simulated timeout or by (2) finding one of those bad edge cases and testing that (but it can be fixed at any time making our test moot) or by (3) overwhelming syntect_server with requests which would be flaky.

My solution tests failure case #2 above and is the only viable way to test it. Your solution would not.

Your solution would be a good way to test the error case #1 above BUT that would be better done as a test inside of the syntect_server codebase (i.e. that it responds with the proper code when a file is too large) AND we haven't yet run into any issues stemming from that error case.

**Comment by slimsag** [comment]
> Can't we use a file that is so enormous it will for sure hit a timeout? I think 3s is very reasonable for an e2e test.

No, this would cause syntect_server to reject the request instead of take a long time processing the request and we would hit a different error path here.

> How does a simulated timeout test a bad language grammar?

It does so by simulating the fact that syntect_server takes a long time to respond (which is what occurs in the case of a bad language grammar).

**Comment by felixfbecker** [comment]
Good points.

> My solution tests failure case #2 above and is the only viable way to test it.

Why is a unit test not viable, if the bug was in frontend logic?

**Comment by slimsag** [comment]
A unit test I think would be great to have in addition, but more generally I think an e2e test is suitable here because:

This didn't stem from "buggy" code from the perspective that the code was unstable/undetermistic or not well written per-say, but rather it appears the issue came from someone just not understanding what we actually want the end-user behavior to be it seems. With a unit test alone, I think we would've had the same issue because the person introducing this regression would've just also updated that test to use this oh-so-lovely new behavior of neglecting timeouts and treating them an error instead of some state we want to communicate.

When I send my 2nd PR in a bit you'll see we've messed this up twice in two completely independent locations and have outright neglected what potential meaning the `aborted` field means and how we want to handle it -- despite the docs existing around it in our GraphQL API. It also has quite bad negative side effects when we do get it wrong, so this all together leads me to believe an e2e test is in order.

This is a bit similar to not having our repository cloning experience be e2e tested, for example, and updating code to just say "error: repository is cloning try again later" calling it a day.

File: pkg/highlight/highlight.go [hunk 1]
```diff
@@ -43,12 +43,15 @@ func IsBinary(content []byte) bool {
...
-func Code(ctx context.Context, content []byte, filepath string, disableTimeout bool, isLightTheme bool) (template.HTML, bool, error) {
+func Code(ctx context.Context, content []byte, filepath string, disableTimeout bool, isLightTheme bool, simulateTimeout bool) (template.HTML, bool, error) {
 	if !disableTimeout {
 		var cancel func()
 		ctx, cancel = context.WithTimeout(ctx, 3*time.Second)
 		defer cancel()
 	}
+	if simulateTimeout {
+		time.Sleep(4 * time.Second)
+	}
```
Comment (keegancsmith) [inline pkg/highlight/highlight.go]: time.Sleep is sadness. Is there no other way to simulate this?

**Comment by felixfbecker** [comment]
> With a unit test alone, I think we would've had the same issue because the person introducing this regression would've just also updated that test to use this oh-so-lovely new behavior of neglecting timeouts and treating them an error instead of some state we want to communicate.

I don't think that's true if there is a test that's literally called "it should still render the unhighlighted blob if highlighting timed out". It's impossible to update that test with the new behavior without contradicting the test description - or put differently, it's just as likely as with an e2e test.

I'm cool with having an e2e test here but we should _at least_ have a unit test. Unit tests are the bottom of the testing pyramid, e2e tests the top, everything that _can_ be unit tested _should_ be unit tested while e2e tests are optional and for important user flows.

File: pkg/highlight/highlight.go (hunk 1 above)
Comment (slimsag) [reply pkg/highlight/highlight.go]: No, there isn't. I know `time.Sleep` is ugly from first glance but it really is the right choice here. I considered adjusting the context timeout on its own and while that would work for this particular issue it would not raise the other UX issues here like the issue this helped me uncover where we apparently also no longer have any loading indicator for files that load slowly :/

**Comment by slimsag** [comment]
@felixfbecker sounds good, I think having both is needed honestly.

**Comment by slimsag** [comment]
@beyang please review
